# **Proyecto 2-A MOS** G27

## Integrantes:
- Juan Esteban Lopez Torres - 202020285
- Bastien Quentin Clement Thirion - 202525085
- Matthieu Maxime Viranin - 202525086
- Alejandro Gonzalez Salazar - 201921465


### Problema fundamental :  
 Optimización de rutas multi-vehículo con restricciones de capacidad, autonomía y operación según el contexto logístico asignado.  

### Justificación :  
 Estos elementos permiten representar adecuadamente la realidad logística del proyecto, ya que cualquier operación de transporte:  
Debe atender clientes con demanda,  
Tiene recursos limitados (vehículos, capacidad, autonomía),   
Y busca minimizar sus costos operativos bajo condiciones específicas del entorno.  
 Sin estos componentes, el modelo no capturaría la complejidad real del sistema ni permitiría comparar soluciones entre diferentes grupos.


## (Caso 1: CVRP Estándar)
### Implementación en Pyomo – Modelo Base (Un solo centro de distribución, vehículos homogéneos)

Implementacion del el **Caso 1 (CVRP estándar)** para LogistiCo:

- Un único centro de distribución (CD01).
- Vehículos homogéneos con la misma capacidad.
- Modelo de **Vehicle Routing Problem con Capacidades (CVRP)** en Pyomo.
- Objetivo: **minimizar el costo operativo**, usando una función de costo basada principalmente en la distancia (extensible a la función objetivo unificada del enunciado).
- Al final, se genera un archivo `verificacion_caso1.csv` con el formato requerido por los lineamientos del proyecto:

`VehicleId, DepotId, InitialLoad, RouteSequence, ClientsServed, DemandsSatisfied, TotalDistance, TotalTime, FuelCost`


In [1]:
# Instalación de Pyomo y del solver GLPK (para Colab)
!pip install pyomo -q
!apt-get install -y glpk-utils > /dev/null


from pyomo.environ import (
    ConcreteModel, Set, Param, Var, NonNegativeReals, Binary,
    Constraint, Objective, SolverFactory, value
)
import math
import csv



In [2]:
# IMPLEMENTACIÓN CVRP ESTÁNDAR (CASO 1)
# Proyecto A

from pyomo.environ import *


# 1. DEFINICIÓN DEL MODELO


model = ConcreteModel()


# 2. CONJUNTOS


# Clientes (IDs estandarizados)
model.N = Set(initialize=[
    "C001","C002","C003","C004","C005","C006","C007","C008","C009","C010",
    "C011","C012","C013","C014","C015","C016","C017","C018","C019","C020",
    "C021","C022","C023","C024"
])

# Vehículos
model.V = Set(initialize=[
    "V001","V002","V003","V004","V005","V006","V007","V008"
])

# Depot
depot = "CD01"

# Nodos (depot + clientes)
model.N0 = Set(initialize=[depot] + list(model.N))

# 3. PARÁMETROS


# Demandas
demand_dict = {
    "C001":13,"C002":15,"C003":12,"C004":15,"C005":20,"C006":17,
    "C007":17,"C008":20,"C009":20,"C010":15,"C011":17,"C012":12,
    "C013":21,"C014":15,"C015":17,"C016":10,"C017":25,"C018":12,
    "C019":11,"C020":15,"C021":14,"C022":18,"C023":15,"C024":11
}
model.demand = Param(model.N, initialize=demand_dict)

# Capacidades de vehículos
capacity_dict = {
    "V001":130,"V002":140,"V003":120,"V004":100,
    "V005":70,"V006":55,"V007":110,"V008":114
}
model.capacity = Param(model.V, initialize=capacity_dict)

# Distancia – Parametrizada genéricamente (rellenada desde OR-Tools)
def dist_placeholder(m, i, j):
    return 0.0
model.dist = Param(model.N0, model.N0, initialize=dist_placeholder, within=NonNegativeReals, mutable=True)

# 4. VARIABLES DE DECISIÓN

# x[i,j,v] = 1 si el vehículo v viaja de i → j
model.x = Var(model.N0, model.N0, model.V, domain=Binary)

# Variables MTZ (subtour elimination)
model.u = Var(model.N, model.V, domain=NonNegativeReals)

# 5. FUNCIÓN OBJETIVO
# Minimización de la distancia total

def objective_rule(m):
    return sum(m.dist[i,j] * m.x[i,j,v] for i in m.N0 for j in m.N0 for v in m.V)
model.OBJ = Objective(rule=objective_rule, sense=minimize)

# 6. RESTRICCIONES DEL CVRP

# No permitir arcos i → i
def nolook_rule(m, i, v):
    return m.x[i,i,v] == 0
model.NoLoops = Constraint(model.N0, model.V, rule=nolook_rule)

# Cada cliente es visitado exactamente una vez
def visit_once_rule(m, j):
    return sum(m.x[i,j,v] for i in m.N0 for v in m.V if i != j) == 1
model.VisitOnce = Constraint(model.N, rule=visit_once_rule)

# Flujo de entrada = flujo de salida para cada cliente y vehículo
def flow_balance_rule(m, h, v):
    return sum(m.x[i,h,v] for i in m.N0 if i != h) == \
           sum(m.x[h,j,v] for j in m.N0 if j != h)
model.FlowBalance = Constraint(model.N, model.V, rule=flow_balance_rule)

# Un vehículo sale del depot a lo sumo una vez
def depot_out_rule(m, v):
    return sum(m.x[depot, j, v] for j in m.N) <= 1
model.DepotOut = Constraint(model.V, rule=depot_out_rule)

# Un vehículo regresa al depot a lo sumo una vez
def depot_in_rule(m, v):
    return sum(m.x[i, depot, v] for i in m.N) <= 1
model.DepotIn = Constraint(model.V, rule=depot_in_rule)

# Balance depot: si sale, debe volver
def depot_balance_rule(m, v):
    return sum(m.x[depot,j,v] for j in m.N) == \
           sum(m.x[i,depot,v] for i in m.N)
model.DepotBalance = Constraint(model.V, rule=depot_balance_rule)

# MTZ: Subtour elimination (solo para i,j clientes)
def mtz_rule(m, i, j, v):
    if i == j:
        return Constraint.Skip
    return m.u[j,v] >= m.u[i,v] + m.demand[j] - m.capacity[v] * (1 - m.x[i,j,v])
model.MTZ = Constraint(model.N, model.N, model.V, rule=mtz_rule)

# Límites de carga u[i,v]
def lower_u(m, i, v):
    return m.u[i,v] >= m.demand[i]
model.ULower = Constraint(model.N, model.V, rule=lower_u)

def upper_u(m, i, v):
    return m.u[i,v] <= m.capacity[v]
model.UUpper = Constraint(model.N, model.V, rule=upper_u)



In [3]:
!pip install ortools folium >/dev/null
from math import radians, sin, cos, atan2, sqrt
from IPython.display import display


from ortools.constraint_solver import pywrapcp, routing_enums_pb2
import math
import csv
import folium


# 1. CASO BASE

# Clientes: (ID estandarizado, latitud, longitud, demanda)
clients_data = [
    ("C001", 4.59795431125545,  -74.09893796560621, 13),
    ("C002", 4.687820646838871, -74.07557103763986, 15),
    ("C003", 4.70949446000624,  -74.10708524062704, 12),
    ("C004", 4.605029068682624, -74.09727965657427, 15),
    ("C005", 4.648463876533332, -74.16464148202755, 20),
    ("C006", 4.662137416953968, -74.12083799988112, 17),
    ("C007", 4.697499030379109, -74.02213076607309, 17),
    ("C008", 4.649416884236942, -74.17207549744595, 20),
    ("C009", 4.606310650273935, -74.15615257246444, 20),
    ("C010", 4.557379705282216, -74.09041145358674, 15),
    ("C011", 4.591594072172954, -74.17802255204528, 17),
    ("C012", 4.7564172406324055,-74.1015410917749,  12),
    ("C013", 4.646217006050524, -74.09690889182339, 21),
    ("C014", 4.725912125314368, -74.1219200708342,  15),
    ("C015", 4.604168478560718, -74.0942948461378,  17),
    ("C016", 4.557320898243896, -74.11138839002187, 10),
    ("C017", 4.615869066082658, -74.12463941285208, 25),
    ("C018", 4.656402930181292, -74.12456164551857, 12),
    ("C019", 4.706188309535041, -74.04990580408855, 11),
    ("C020", 4.746624762561149, -74.121866801394,   15),
    ("C021", 4.746517184687941, -74.02454895778163, 14),
    ("C022", 4.625425626584072, -74.08683592702927, 18),
    ("C023", 4.731406773645735, -74.11652648751144, 15),
    ("C024", 4.577489966973782, -74.12629327283491, 11),
]

# Depot único (CD01)
depot_id = "CD01"
depot_lat = 4.743359
depot_lon = -74.153536

# Vehículos: (ID, capacidad, rango/autonomía en km)
vehicles_data = [
    ("V001", 130, 170),
    ("V002", 140, 200),
    ("V003", 120, 180),
    ("V004", 100,  90),
    ("V005",  70, 100),
    ("V006",  55, 170),
    ("V007", 110, 150),
    ("V008", 114, 140),
]

# Parámetros base de combustible (para el archivo de verificación)
fuel_price = 16300            # COP/gallon
fuel_efficiency = 30          # km/gallon, típica
speed_kmh = 30                # Velocidad promedio urbana (para estimar tiempo)

# 2. PREPARAR NODOS Y MATRIZ DE DISTANCIAS

# Lista de nodos estandarizados: primero el depot, luego clientes
node_ids = [depot_id] + [c[0] for c in clients_data]

# Diccionarios de coordenadas
lat = {depot_id: depot_lat}
lon = {depot_id: depot_lon}
demand = {}

for cid, la, lo, dem in clients_data:
    lat[cid] = la
    lon[cid] = lo
    demand[cid] = dem

# Distancia Haversine en km
def haversine_km(lat1, lon1, lat2, lon2):
    R = 6371.0
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlambda = math.radians(lon2 - lon1)
    a = math.sin(dphi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlambda / 2) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c

# Construir matriz de distancias (en km)
num_nodes = len(node_ids)
distance_km = [[0.0 for _ in range(num_nodes)] for _ in range(num_nodes)]

for i in range(num_nodes):
    for j in range(num_nodes):
        if i == j:
            distance_km[i][j] = 0.0
        else:
            ni = node_ids[i]
            nj = node_ids[j]
            distance_km[i][j] = haversine_km(lat[ni], lon[ni], lat[nj], lon[nj])

# OR-Tools trabaja mejor con enteros, escalamos a metros como enteros
distance_m = [[int(distance_km[i][j] * 1000) for j in range(num_nodes)] for i in range(num_nodes)]

# Demanda por nodo (0 para el depot)
demands = [0]  # depot
demands.extend([demand[c[0]] for c in clients_data])

# Capacidades por vehículo
vehicle_ids = [v[0] for v in vehicles_data]
vehicle_capacities = [v[1] for v in vehicles_data]
num_vehicles = len(vehicle_ids)
depot_index = 0  # índice del depot en node_ids

# 3. CONSTRUIR Y RESOLVER MODELO CVRP CON OR-TOOLS

# Manager de índices
manager = pywrapcp.RoutingIndexManager(num_nodes, num_vehicles, depot_index)

# Modelo de ruteo
routing = pywrapcp.RoutingModel(manager)

# Callback de distancia
def distance_callback(from_index, to_index):
    from_node = manager.IndexToNode(from_index)
    to_node = manager.IndexToNode(to_index)
    return distance_m[from_node][to_node]

transit_callback_index = routing.RegisterTransitCallback(distance_callback)

# Costo: distancia recorrida
routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

# Dimensión de capacidad
def demand_callback(from_index):
    from_node = manager.IndexToNode(from_index)
    return demands[from_node]

demand_callback_index = routing.RegisterUnaryTransitCallback(demand_callback)
routing.AddDimensionWithVehicleCapacity(
    demand_callback_index,
    0,  # no slack
    vehicle_capacities,
    True,  # start cumul at zero
    "Capacity",
)

# Parámetros de búsqueda
search_parameters = pywrapcp.DefaultRoutingSearchParameters()
search_parameters.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC
search_parameters.local_search_metaheuristic = routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH
search_parameters.time_limit.FromSeconds(30)  # límite de tiempo razonable

solution = routing.SolveWithParameters(search_parameters)

if not solution:
    print("No se encontró solución factible :(")
else:
    print("Solución encontrada.")
    # Costo total en metros y en km
    total_distance_m = 0
    for v in range(num_vehicles):
        index = routing.Start(v)
        while not routing.IsEnd(index):
            previous_index = index
            index = solution.Value(routing.NextVar(index))
            total_distance_m += routing.GetArcCostForVehicle(previous_index, index, v)
    print(f"Distancia total (km): {total_distance_m / 1000:.4f}")

# 4. RECONSTRUIR RUTAS Y GENERAR verificacion_caso1.csv

filename = "verificacion_caso1.csv"
if solution:
    with open(filename, mode="w", newline="") as f:
        writer = csv.writer(f)
        # Encabezado EXACTO
        writer.writerow([
            "VehicleId", "DepotId", "InitialLoad", "RouteSequence",
            "ClientsServed", "DemandsSatisfied", "TotalDistance",
            "TotalTime", "FuelCost"
        ])

        for v in range(num_vehicles):
            index = routing.Start(v)
            route_nodes_idx = []
            route_nodes_ids = []

            # Construimos la ruta en términos de índices y IDs
            while not routing.IsEnd(index):
                node_index = manager.IndexToNode(index)
                route_nodes_idx.append(node_index)
                route_nodes_ids.append(node_ids[node_index])
                index = solution.Value(routing.NextVar(index))
            # Añadir nodo final (depot)
            node_index = manager.IndexToNode(index)
            route_nodes_idx.append(node_index)
            route_nodes_ids.append(node_ids[node_index])

            # Clientes servidos (excluyendo depot)
            clients_in_route = [nid for nid in route_nodes_ids if nid != depot_id]
            clients_served = len(clients_in_route)

            # Si el vehículo no sirve ningún cliente, lo omitimos del archivo
            if clients_served == 0:
                continue

            # Ruta como string CD01-Cxxx-...-CD01
            route_sequence = "-".join(route_nodes_ids)

            # Demandas satisfechas en orden
            demands_list = [demand[cid] for cid in clients_in_route]
            demands_satisfied_str = "-".join(str(d) for d in demands_list)

            # Carga inicial
            initial_load = sum(demands_list)

            # Distancia total en km usando la matriz real distance_km
            total_distance = 0.0
            for i in range(len(route_nodes_idx) - 1):
                ni = route_nodes_idx[i]
                nj = route_nodes_idx[i + 1]
                total_distance += distance_km[ni][nj]

            # Tiempo total (minutos)
            total_time_minutes = (total_distance / speed_kmh) * 60.0

            # Costo de combustible
            gallons_used = total_distance / fuel_efficiency
            fuel_cost = gallons_used * fuel_price

            writer.writerow([
                vehicle_ids[v],               # VehicleId
                depot_id,                     # DepotId
                round(initial_load, 2),       # InitialLoad
                route_sequence,               # RouteSequence
                clients_served,               # ClientsServed
                demands_satisfied_str,        # DemandsSatisfied
                round(total_distance, 4),     # TotalDistance (km)
                round(total_time_minutes, 2), # TotalTime (min)
                round(fuel_cost, 2),          # FuelCost (COP)
            ])

    print(f"Archivo '{filename}' generado correctamente.")

# 5. MAPA INTERACTIVO CON FOLIUM (MOSTRAR EN COLAB)

from IPython.display import display

if solution:
    m = folium.Map(location=[depot_lat, depot_lon], zoom_start=11)

    # Marcador del depot
    folium.Marker(
        location=[depot_lat, depot_lon],
        popup=f"Depot {depot_id}",
        icon=folium.Icon(color="red", icon="home")
    ).add_to(m)

    # Marcadores de clientes
    for cid, la, lo, dem in clients_data:
        folium.CircleMarker(
            location=[la, lo],
            radius=4,
            popup=f"{cid} - Demanda: {dem}",
        ).add_to(m)

    # Colores para rutas
    colors = [
        "blue", "green", "purple", "orange",
        "darkred", "cadetblue", "darkgreen", "black"
    ]

    for v in range(num_vehicles):
        index = routing.Start(v)
        route_nodes_ids = []

        while not routing.IsEnd(index):
            node_index = manager.IndexToNode(index)
            route_nodes_ids.append(node_ids[node_index])
            index = solution.Value(routing.NextVar(index))
        # nodo final
        node_index = manager.IndexToNode(index)
        route_nodes_ids.append(node_ids[node_index])

        # Clientes servidos
        clients_in_route = [nid for nid in route_nodes_ids if nid != depot_id]
        if len(clients_in_route) == 0:
            continue

        latlon_route = [(lat[nid], lon[nid]) for nid in route_nodes_ids]
        folium.PolyLine(
            locations=latlon_route,
            popup=f"Ruta {vehicle_ids[v]}",
            weight=4,
            opacity=0.8,
            color=colors[v % len(colors)],
        ).add_to(m)

    #  Mostrar el mapa directo en pantalla
    display(m)


Solución encontrada.
Distancia total (km): 124.9390
Archivo 'verificacion_caso1.csv' generado correctamente.


La configuración inicial del modelo de ruteo permitió implementar y resolver el Problema de Rutas de Vehículos con capacidad (CVRP) en un entorno urbano realista dentro de Bogotá. A partir de los datos georreferenciados de 24 clientes y un centro de distribución, el sistema de optimización  generó rutas factibles respetando las capacidades máximas de la flota y minimizando la distancia total recorrida.

Los resultados obtenidos son los siguientes:

- **Distancia total recorrida:** 124.939 km
- **Archivo de validación generado:** `verificacion_caso1.csv`  
  → Este archivo permite rastrear:
  - Clientes atendidos por cada vehículo  
  - Demandas satisfechas  
  - Secuencia completa de la ruta (CD → clientes → CD)  
  - Distancia real por vehículo  
  - Tiempo estimado  
  - Costo de combustible

La metodología siguió los lineamientos establecidos para el Caso Base:

1. **Uso de distancias reales (Haversine):**  
2. **Restricciones de capacidad:**  
3. **Minimización del recorrido total:**  
4. **Visualización geográfica:**  


En resumen, el modelo logró planificar rutas factibles, minimizando la distancia y cumpliendo las restricciones operativas básicas, lo cual sienta las bases para las siguientes fases del proyecto donde se incorporarán centros múltiples, ventanas de tiempo y condiciones urbanas específicas.


## Caso 2: Logística Urbana con Múltiples Depósitos (MDVRP)

### Evolución del Modelo
A diferencia del Caso Base (CVRP estándar con un único origen), este escenario aborda la complejidad real de la operación en Bogotá introduciendo la gestión de **Múltiples Centros de Distribución (CDs)**.

El desafío principal deja de ser solo el ruteo y pasa a ser la **asignación estratégica**: decidir qué vehículo sale de qué depósito para minimizar costos sin violar las restricciones de inventario.

### Enfoque de Resolución
Para cumplir con los requisitos de implementación formal, desarrollamos la solución a **Pyomo** utilizando un enfoque de **Programación Lineal Entera Mixta (MILP)** con las siguientes características avanzadas:

1.  **Modelo MDVRP (Multi-Depot VRP):** Se introducen variables binarias de decisión $z_{k,d}$ para asignar dinámicamente cada vehículo a un depósito óptimo.
2.  **Restricciones de Capacidad Robustas:** Se modela el inventario de cada depósito. Para evitar la infactibilidad matemática en casos donde la demanda local supera la capacidad instalada (como se advierte en el *README*), se implementan **variables de holgura (slack)** con penalización en la función objetivo. Esto permite al modelo reportar déficits estructurales en lugar de fallar.
3.  **Solver Exacto:** Se utiliza **CBC (Coin-OR)** con un tiempo extendido de búsqueda para garantizar la convergencia al óptimo global o la mejor solución factible posible.

---

In [4]:
# ==========================================
# CELDA 1: INSTALACIÓN DE DEPENDENCIAS
# ==========================================
import sys
import os

# Detectar si estamos en Google Colab o Linux para instalar CBC
if 'google.colab' in sys.modules or os.name == 'posix':
    print("Instalando COIN-OR CBC y librerías necesarias...")
    !apt-get install -y -qq coinor-cbc
    !pip install -q pyomo pandas folium matplotlib
else:
    print("Estás en Windows/Mac local. Asegúrate de tener el binario 'cbc' en tu PATH.")

print("Entorno configurado con CBC.")

Instalando COIN-OR CBC y librerías necesarias...
Entorno configurado con CBC.


In [5]:
# ==========================================
# CELDA 2: IMPORTACIONES Y DATOS DE ENTRADA
# ==========================================
import pandas as pd
import pyomo.environ as pyo
from pyomo.opt import SolverFactory, TerminationCondition
import folium
import math
from IPython.display import display

# --- 1. PARÁMETROS ECONÓMICOS (parameters_urban.csv) ---
PARAMS = {
    'COST_FIXED': 50000,      # COP/vehículo
    'COST_DIST': 2500,        # COP/km
    'COST_TIME': 7600,        # COP/hora
    'FUEL_PRICE': 16300,      # COP/galón
    'AVG_SPEED': 25           # km/h
}

# Eficiencias calculadas (Promedio Min/Max)
EFFICIENCY = {
    'small van': 40.0,    # (35+45)/2
    'medium van': 30.0,   # (25+35)/2
    'light truck': 25.0   # (22+28)/2
}

# --- 2. CLIENTES (clients.csv) ---
# Latitudes y Longitudes
clients_data = [
    {'id': 'C001', 'lat': 4.63255284, 'lon': -74.19699185, 'dem': 12},
    {'id': 'C002', 'lat': 4.60132821, 'lon': -74.15503659, 'dem': 15},
    {'id': 'C003', 'lat': 4.73242104, 'lon': -74.10178732, 'dem': 15},
    {'id': 'C004', 'lat': 4.63861219, 'lon': -74.19486224, 'dem': 6},
    {'id': 'C005', 'lat': 4.72769161, 'lon': -74.11027242, 'dem': 5},
    {'id': 'C006', 'lat': 4.66500293, 'lon': -74.15228947, 'dem': 11},
    {'id': 'C007', 'lat': 4.67710170, 'lon': -74.03241076, 'dem': 12},
    {'id': 'C008', 'lat': 4.70700701, 'lon': -74.06247587, 'dem': 10},
    {'id': 'C009', 'lat': 4.63607455, 'lon': -74.09804169, 'dem': 15}
]

# --- 3. DEPÓSITOS (depots.csv) ---
depots_data = [
    {'id': 'CD01', 'lat': 4.75021191, 'lon': -74.08124218, 'cap': 8},
    {'id': 'CD02', 'lat': 4.53638322, 'lon': -74.10993359, 'cap': 10},
    {'id': 'CD03', 'lat': 4.79292596, 'lon': -74.03854815, 'cap': 0},
    {'id': 'CD04', 'lat': 4.72167778, 'lon': -74.06706883, 'cap': 4},
    {'id': 'CD05', 'lat': 4.60770705, 'lon': -74.13826338, 'cap': 28},
    {'id': 'CD06', 'lat': 4.65046305, 'lon': -74.12400186, 'cap': 3},
    {'id': 'CD07', 'lat': 4.62191177, 'lon': -74.09561875, 'cap': 0},
    {'id': 'CD08', 'lat': 4.67896068, 'lon': -74.10975624, 'cap': 10},
    {'id': 'CD09', 'lat': 4.73597306, 'lon': -74.09547236, 'cap': 43},
    {'id': 'CD10', 'lat': 4.55064099, 'lon': -74.10991610, 'cap': 1},
    {'id': 'CD11', 'lat': 4.66470296, 'lon': -74.10977422, 'cap': 16},
    {'id': 'CD12', 'lat': 4.57917406, 'lon': -74.12408926, 'cap': 18}
]

# --- 4. VEHÍCULOS (vehicles.csv) ---
vehicles_data = [
    {'id': 'V001', 'type': 'small van', 'cap': 131.92, 'range': 145.85},
    {'id': 'V002', 'type': 'medium van', 'cap': 108.44, 'range': 1304.61},
    {'id': 'V003', 'type': 'medium van', 'cap': 91.50, 'range': 953.17},
    {'id': 'V004', 'type': 'light truck', 'cap': 32.90, 'range': 17.30},
    {'id': 'V005', 'type': 'medium van', 'cap': 22.65, 'range': 16.63},
    {'id': 'V006', 'type': 'light truck', 'cap': 22.68, 'range': 13.60}
]

# --- ESTRUCTURACIÓN DE DATOS ---
nodes = {}
for d in depots_data:
    nodes[d['id']] = {'lat': d['lat'], 'lon': d['lon'], 'cap': d['cap'], 'dem': 0, 'type': 'depot'}
for c in clients_data:
    nodes[c['id']] = {'lat': c['lat'], 'lon': c['lon'], 'dem': c['dem'], 'cap': 0, 'type': 'client'}

vehicles = {v['id']: v for v in vehicles_data}

all_nodes = [d['id'] for d in depots_data] + [c['id'] for c in clients_data]
client_ids = [c['id'] for c in clients_data]
depot_ids = [d['id'] for d in depots_data]
vehicle_ids = [v['id'] for v in vehicles_data]

print(" Datos validados y cargados correctamente.")
print(f"   - {len(clients_data)} Clientes")
print(f"   - {len(depots_data)} Depósitos")
print(f"   - {len(vehicles_data)} Vehículos")

 Datos validados y cargados correctamente.
   - 9 Clientes
   - 12 Depósitos
   - 6 Vehículos


### Formulación Matemática y Estrategia de Robustez

Para garantizar la operatividad del modelo ante restricciones de inventario estrictas, se implementó una formulación **robusta** que penaliza pero permite violaciones controladas de capacidad.

**Función Objetivo Extendida:**
El objetivo es minimizar el Costo Total Generalizado ($Z$), compuesto por costos operativos reales y penalizaciones virtuales:

$$\min Z = \sum (C_{fijo} + C_{dist} + C_{tiempo} + C_{combustible}) + \sum_{d \in D} (Slack_d \times P)$$

Donde:
* **Costos Operativos:** Suma de activación de flota, distancia recorrida, tiempo de conductor y consumo de gasolina según eficiencia del vehículo.
* **$Slack_d$:** Variable continua no negativa que representa el exceso de carga (kg) en el depósito $d$.
* **$P$:** Penalización ("Big M"), configurada en $10,000,000$ COP/kg.

**Justificación:**
Esta estructura asegura que el solver priorice siempre el cumplimiento de capacidades ($Slack = 0$). Solo en casos donde la demanda local supere físicamente la capacidad instalada (infactibilidad estructural), el modelo activará la variable de holgura, permitiendo identificar y cuantificar los **cuellos de botella** en la infraestructura.

In [6]:
# ==========================================
# CELDA 3: MODELO PYOMO OPTIMIZADO
# ==========================================

# 1. Distancias (Haversine)
def haversine(lat1, lon1, lat2, lon2):
    R = 6371
    dlat, dlon = math.radians(lat2 - lat1), math.radians(lon2 - lon1)
    a = math.sin(dlat/2)**2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon/2)**2
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))

dist_matrix = {}
time_matrix = {}
for i in all_nodes:
    for j in all_nodes:
        d = haversine(nodes[i]['lat'], nodes[i]['lon'], nodes[j]['lat'], nodes[j]['lon'])
        dist_matrix[i, j] = d
        time_matrix[i, j] = d / PARAMS['AVG_SPEED']

# 2. Inicializar Modelo
m = pyo.ConcreteModel()
m.V = pyo.Set(initialize=vehicle_ids)
m.N = pyo.Set(initialize=client_ids)
m.D = pyo.Set(initialize=depot_ids)
m.Nodes = pyo.Set(initialize=all_nodes)

# Variables
m.x = pyo.Var(m.Nodes, m.Nodes, m.V, domain=pyo.Binary)
m.z = pyo.Var(m.V, m.D, domain=pyo.Binary)
m.u = pyo.Var(m.N, m.V, domain=pyo.NonNegativeReals)
m.q = pyo.Var(m.V, m.D, domain=pyo.NonNegativeReals) # Carga asignada
m.slack_depot = pyo.Var(m.D, domain=pyo.NonNegativeReals) # Holgura de capacidad

# Función Objetivo (Con Penalización de 10 Millones para forzar cumplimiento)
PENALTY = 10000000

def obj_rule(model):
    c_dist = sum(dist_matrix[i,j] * model.x[i,j,k] * PARAMS['COST_DIST'] for i in model.Nodes for j in model.Nodes for k in model.V)
    c_time = sum(time_matrix[i,j] * model.x[i,j,k] * PARAMS['COST_TIME'] for i in model.Nodes for j in model.Nodes for k in model.V)
    c_fixed = sum(model.z[k,d] * PARAMS['COST_FIXED'] for k in model.V for d in model.D)
    c_fuel = 0
    for k in model.V:
        eff = EFFICIENCY.get(vehicles[k]['type'], 30.0)
        c_fuel += sum(dist_matrix[i,j] * model.x[i,j,k] for i in model.Nodes for j in model.Nodes) / eff * PARAMS['FUEL_PRICE']

    # Penalización por uso de holgura
    c_penalty = sum(model.slack_depot[d] for d in model.D) * PENALTY
    return c_dist + c_time + c_fixed + c_fuel + c_penalty

m.Obj = pyo.Objective(rule=obj_rule, sense=pyo.minimize)

# Restricciones
m.Visit = pyo.Constraint(m.N, rule=lambda mod, i: sum(mod.x[i,j,k] for j in mod.Nodes for k in mod.V if i!=j) == 1)
m.Flow = pyo.Constraint(m.N, m.V, rule=lambda mod, h, k: sum(mod.x[i,h,k] for i in mod.Nodes if i!=h) == sum(mod.x[h,j,k] for j in mod.Nodes if j!=h))
m.OneDepot = pyo.Constraint(m.V, rule=lambda mod, k: sum(mod.z[k,d] for d in mod.D) <= 1)
m.StartAtDepot = pyo.Constraint(m.V, m.D, rule=lambda mod, k, d: sum(mod.x[d,j,k] for j in mod.N) == mod.z[k,d])
m.EndAtDepot = pyo.Constraint(m.V, m.D, rule=lambda mod, k, d: sum(mod.x[i,d,k] for i in mod.N) == mod.z[k,d])
m.CapVehicle = pyo.Constraint(m.V, rule=lambda mod, k: sum(nodes[i]['dem'] * sum(mod.x[i,j,k] for j in mod.Nodes if i!=j) for i in mod.N) <= vehicles[k]['cap'])
m.RangeVehicle = pyo.Constraint(m.V, rule=lambda mod, k: sum(dist_matrix[i,j] * mod.x[i,j,k] for i in mod.Nodes for j in mod.Nodes) <= vehicles[k]['range'])

# Restricciones de Capacidad de Depósito
m.LoadBalance = pyo.Constraint(m.V, rule=lambda mod, k: sum(mod.q[k,d] for d in mod.D) == sum(nodes[i]['dem'] * sum(mod.x[i,j,k] for j in mod.Nodes if i!=j) for i in mod.N))
m.LoadUpperBound = pyo.Constraint(m.V, m.D, rule=lambda mod, k, d: mod.q[k,d] <= vehicles[k]['cap'] * mod.z[k,d])
m.DepotCap = pyo.Constraint(m.D, rule=lambda mod, d: sum(mod.q[k,d] for k in mod.V) <= nodes[d]['cap'] + mod.slack_depot[d])
m.MTZ = pyo.Constraint(m.N, m.N, m.V, rule=lambda mod, i, j, k: mod.u[i,k] - mod.u[j,k] + len(mod.N)*mod.x[i,j,k] <= len(mod.N)-1 if i!=j else pyo.Constraint.Skip)

print("Modelo construido.")

Modelo construido.


In [7]:
# ==========================================
# CELDA 4: SOLVER CBC Y GENERACIÓN DE DATOS
# ==========================================
import pandas as pd

# 1. Configuración y Resolución
solver = SolverFactory('cbc')
solver.options['seconds'] = 600

print(f" Optimizando con CBC (Max {solver.options['seconds']}s)...")
results = solver.solve(m)

term_cond = results.solver.termination_condition
print(f"\n--- ESTADO: {term_cond} ---")

# 2. Procesamiento de Resultados
if term_cond in [TerminationCondition.optimal, TerminationCondition.maxTimeLimit, TerminationCondition.feasible]:
    output_data = []

    for k in m.V:
        # A. Encontrar depósito de inicio
        used_depot = None
        for d in m.D:
            if pyo.value(m.z[k,d]) > 0.5:
                used_depot = d
                break

        if used_depot:
            # B. Reconstrucción de Ruta
            route = [used_depot]
            curr = used_depot
            dist_total = 0

            # Límite de seguridad para evitar bucles infinitos
            for step in range(len(all_nodes) + 5):
                next_node = None

                # Buscar el siguiente nodo conectado (x=1)
                for j in m.Nodes:
                    if j != curr:
                        if pyo.value(m.x[curr, j, k]) > 0.5:
                            next_node = j
                            break

                # Si encontramos el siguiente nodo:
                if next_node:
                    dist_segment = dist_matrix[curr, next_node]
                    dist_total += dist_segment
                    route.append(next_node)
                    curr = next_node

                    # CONDICIÓN DE PARADA: Si volvimos a un depósito
                    if curr in depot_ids:
                        break
                else:
                    # Ruta incompleta o rota
                    break

            # C. Calcular Métricas
            # Filtrar solo clientes para sumar demanda
            client_nodes = [n for n in route if n in client_ids]
            initial_load = sum(nodes[n]['dem'] for n in client_nodes)

            # Eficiencia
            v_type = vehicles[k]['type']
            eff = EFFICIENCY.get(v_type, 30.0)

            # Costos
            fuel_cost = (dist_total / eff) * PARAMS['FUEL_PRICE']
            time_total = (dist_total / PARAMS['AVG_SPEED']) * 60 # Minutos

            output_data.append({
                'VehicleId': k,
                'DepotId': used_depot,
                'InitialLoad': initial_load,
                'RouteSequence': "-".join(route),
                'ClientsServed': len(client_nodes),
                'DemandsSatisfied': "-".join([str(int(nodes[n]['dem'])) for n in client_nodes]),
                'TotalDistance': round(dist_total, 2),
                'TotalTime': round(time_total, 2),
                'FuelCost': round(fuel_cost, 0)
            })

    # 3. Generar DataFrame y CSV
    df_out = pd.DataFrame(output_data)

    # Asegurar que las columnas numéricas sean reconocidas como números
    cols_num = ['TotalDistance', 'TotalTime', 'FuelCost', 'InitialLoad', 'ClientsServed']
    for c in cols_num:
        if c in df_out.columns:
            df_out[c] = pd.to_numeric(df_out[c])

    df_out.to_csv('verificacion_caso2.csv', index=False)
    print(" Solución procesada. Archivo 'verificacion_caso2.csv' generado correctamente.")

    # Mostrar vista previa para verificar que no haya ceros
    try:
        from IPython.display import display
        display(df_out.head())
    except:
        print(df_out.head())

else:
    print(" No se encontró solución.")

 Optimizando con CBC (Max 600s)...

--- ESTADO: optimal ---
 Solución procesada. Archivo 'verificacion_caso2.csv' generado correctamente.


VehicleId DepotId  InitialLoad                  RouteSequence  \
0      V001    CD09           42  CD09-C008-C007-C005-C003-CD09   
1      V002    CD12           18            CD12-C001-C004-CD12   
2      V003    CD05           26            CD05-C002-C006-CD05   
3      V005    CD11           15                 CD11-C009-CD11   

   ClientsServed DemandsSatisfied  TotalDistance  TotalTime  FuelCost  
0              4       10-12-5-15          21.76      52.23    8868.0  
1              2             12-6          21.00      50.39   11408.0  
2              2            15-11          15.63      37.52    8495.0  
3              1               15           6.88      16.51    3737.0

### Análisis de Resultados y Visualización Geoespacial

A continuación, se procesan los resultados de la optimización para generar inteligencia de negocio accionable.

**Estructura del Análisis:**
1.  **KPIs Financieros:** Se desglosa el costo en sus componentes reales (dinero a desembolsar) vs. costos matemáticos (multas).
    * *Nota:* Si el "Costo Matemático" es superior al "Real", indica que existen depósitos operando por encima de su capacidad nominal.
2.  **Estado de la Infraestructura:** Tabla detallada de utilización de cada Centro de Distribución (CD).
    * 🟢 **OK:** Operación normal.
    * 🔴 **SOBRECUPO:** El depósito excede su capacidad (cuello de botella identificado).
    * ⚪ **SIN USO:** Depósito no asignado por ineficiencia o lejanía.
3.  **Visualización de Rutas (Folium):**
    * Las rutas se colorean según el **Depósito de Origen** para validar visualmente la clusterización geográfica.
    * Los marcadores de depósito incluyen alertas visuales si presentan sobrecupo.

In [8]:
# ==========================================
# CELDA 5: REPORTE GERENCIAL
# ==========================================
if 'df_out' in locals() and not df_out.empty:
    print("\n" + "="*60)
    print("          REPORTE - CASO 2 (CBC)")
    print("="*60)

    # Asegurar tipos numéricos
    df_out['TotalDistance'] = pd.to_numeric(df_out['TotalDistance'])
    df_out['TotalTime'] = pd.to_numeric(df_out['TotalTime'])
    df_out['FuelCost'] = pd.to_numeric(df_out['FuelCost'])

    # Cálculo de Costos
    total_dist = df_out['TotalDistance'].sum()
    cost_dist = total_dist * PARAMS['COST_DIST']
    cost_time = (df_out['TotalTime'].sum() / 60) * PARAMS['COST_TIME']
    cost_fuel = df_out['FuelCost'].sum()
    cost_fixed = len(df_out) * PARAMS['COST_FIXED']

    total_operation = cost_dist + cost_time + cost_fuel + cost_fixed

    # Penalizaciones Teóricas
    slack_cost = 0
    if hasattr(m, 'slack_depot'):
        slack_cost = sum(pyo.value(m.slack_depot[d]) for d in m.D) * PENALTY

    print(f"1. RESUMEN FINANCIERO")
    print(f"   ------------------------------------------------")
    print(f"   COSTO OPERATIVO REAL:      ${total_operation:,.0f} COP")
    if slack_cost > 0.1: # Umbral bajo por si hay decimales
        print(f"   COSTO MATEMÁTICO (+MULTA): ${total_operation + slack_cost:,.0f} COP")
    print(f"   ------------------------------------------------")
    print(f"   - Flota (Fijo):            ${cost_fixed:,.0f} ({len(df_out)} vehículos)")
    print(f"   - Distancia Variable:      ${cost_dist:,.0f} ({total_dist:.1f} km)")
    print(f"   - Tiempo:                  ${cost_time:,.0f}")
    print(f"   - Combustible:             ${cost_fuel:,.0f}")

    print(f"\n2. ESTADO DE LOS DEPÓSITOS")
    print(f"   ------------------------------------------------")
    print(f"   {'Depósito':<10} | {'Capacidad':<10} | {'Uso':<10} | {'Estado':<15}")

    for d_id in depot_ids:
        usage = sum(pyo.value(m.q[k,d_id]) for k in m.V)
        cap = nodes[d_id]['cap']
        slack = pyo.value(m.slack_depot[d_id])

        status = "🟢 OK"
        if slack > 0.1: status = f"🔴 SOBRECUPO (+{slack:.1f})"
        elif usage == 0: status = "⚪ SIN USO"

        if usage > 0 or slack > 0.1:
            print(f"   {d_id:<10} | {cap:<10.0f} | {usage:<10.0f} | {status:<15}")
    print("="*60)
else:
    print(" No hay datos para reportar.")


          REPORTE - CASO 2 (CBC)
1. RESUMEN FINANCIERO
   ------------------------------------------------
   COSTO OPERATIVO REAL:      $415,525 COP
   ------------------------------------------------
   - Flota (Fijo):            $200,000 (4 vehículos)
   - Distancia Variable:      $163,175 (65.3 km)
   - Tiempo:                  $19,842
   - Combustible:             $32,508

2. ESTADO DE LOS DEPÓSITOS
   ------------------------------------------------
   Depósito   | Capacidad  | Uso        | Estado         
   CD05       | 28         | 26         | 🟢 OK           
   CD09       | 43         | 42         | 🟢 OK           
   CD11       | 16         | 15         | 🟢 OK           
   CD12       | 18         | 18         | 🟢 OK           


In [9]:
# ==========================================
# CELDA 6: MAPA INTERACTIVO
# ==========================================
import folium

# Verificar si hay datos para graficar
if 'df_out' in locals() and not df_out.empty:
    print("Iniciando generación del mapa...")

    # 1. Configurar Mapa Base (Centrado usando clients_data)
    # Usamos la lista clients_data que definimos en la Celda 2
    try:
        avg_lat = sum(c['lat'] for c in clients_data) / len(clients_data)
        avg_lon = sum(c['lon'] for c in clients_data) / len(clients_data)
    except:
        avg_lat, avg_lon = 4.65, -74.1 # Fallback a Bogotá centro si falla el cálculo

    mapa = folium.Map([avg_lat, avg_lon], zoom_start=12, tiles='CartoDB positron')

    # 2. Colores
    colors = ['red', 'blue', 'green', 'purple', 'orange', 'darkred', 'cadetblue', 'black']

    # 3. Dibujar Depósitos
    print("   - Dibujando depósitos...")
    for d in depot_ids:
        node = nodes[d] # Buscar en el diccionario de nodos

        # Verificar holgura
        slack_val = 0
        if hasattr(m, 'slack_depot'):
            slack_val = pyo.value(m.slack_depot[d])

        popup_txt = f"{d}<br>Cap: {node['cap']}"
        icon_color = 'black'

        if slack_val > 0.1:
            popup_txt += f"<br>SOBRECUPO: {slack_val:.1f}"
            icon_color = 'red'

        folium.Marker(
            [node['lat'], node['lon']],
            popup=popup_txt,
            icon=folium.Icon(color=icon_color, icon='warehouse', prefix='fa')
        ).add_to(mapa)

    # 4. Dibujar Clientes
    print("   - Dibujando clientes...")
    for c in client_ids:
        node = nodes[c]
        folium.CircleMarker(
            [node['lat'], node['lon']],
            radius=5, color='gray', fill=True,
            popup=f"{c} ({node['dem']}kg)"
        ).add_to(mapa)

    # 5. Dibujar Rutas (CON DIAGNÓSTICO)
    print("   - Dibujando rutas...")
    count_routes = 0

    for _, r in df_out.iterrows():
        # Obtener secuencia y limpiar espacios por si acaso
        seq_str = str(r['RouteSequence']).strip()
        node_list = seq_str.split('-')

        # Construir lista de coordenadas
        pts = []
        for n in node_list:
            n_clean = n.strip()
            if n_clean in nodes:
                pts.append([nodes[n_clean]['lat'], nodes[n_clean]['lon']])
            else:
                print(f"Advertencia: Nodo {n_clean} no encontrado en coordenadas.")

        if len(pts) > 1:
            # Elegir color basado en el depósito
            try:
                color_idx = depot_ids.index(r['DepotId']) % len(colors)
                color = colors[color_idx]
            except:
                color = 'blue'

            folium.PolyLine(
                pts,
                color=color,
                weight=3,
                opacity=0.8,
                tooltip=f"{r['VehicleId']} ({r['DepotId']})"
            ).add_to(mapa)
            count_routes += 1

    print(f" Se dibujaron {count_routes} rutas exitosamente.")

    # 6. Guardar y Mostrar
    filename = "mapa_final_cbc.html"
    mapa.save(filename)
    print(f" Archivo guardado: {filename}")

    try:
        from IPython.display import display
        display(mapa)
    except:
        print("Abre el archivo HTML para ver el mapa.")

else:
    print(" Error: No se encontró el DataFrame 'df_out' con los resultados.")

Iniciando generación del mapa...
   - Dibujando depósitos...
   - Dibujando clientes...
   - Dibujando rutas...
 Se dibujaron 4 rutas exitosamente.
 Archivo guardado: mapa_final_cbc.html


# Análisis de Resultados y Conclusiones Finales (Caso 2)

### Desempeño del Modelo
La implementación robusta en **Pyomo** utilizando el solver **CBC** fue exitosa, logrando converger a una **Solución Óptima Global** en el tiempo asignado.
* **Estado Final:** `Optimal` (Gap 0.00%).
* **Validación:** Se cumplieron el 100% de las demandas de los clientes sin violar restricciones duras de capacidad, evitando el pago de penalizaciones.

### Análisis Financiero
El costo operativo total minimizado es de $415,525 COP. La estructura de costos revela la naturaleza de la operación:

1. Costos Fijos (48%): $200,000 COP. Representa el mayor componente,
 derivado de la activación de 4 vehículos. Esto sugiere que la reducción de flota es la palanca más fuerte para ahorro futuro.

2. Costos Variables (Distancia + Tiempo + Combustible): ~$215,000 COP. La optimización de rutas logró comprimir el recorrido total a solo **65.3 km** para atender toda la demanda, una cifra altamente eficiente para la dispersión geográfica de los clientes.

### Gestión de Infraestructura y Cuellos de Botella
El análisis de uso de los Centros de Distribución (CDs) arroja hallazgos estratégicos críticos:

| Depósito | Capacidad | Uso Real | % Ocupación | Diagnóstico |
| :--- | :--- | :--- | :--- | :--- |
| **CD12** | 18 kg | 18 kg | **100%** | 🔴 **Cuello de Botella Crítico** |
| **CD09** | 43 kg | 42 kg | **98%** | 🟡 **Saturado** |
| **CD05** | 28 kg | 26 kg | **93%** | 🟢 Eficiente |
| **CD11** | 16 kg | 15 kg | **94%** | 🟢 Eficiente |
| **CD06** | 3 kg | 0 kg | 0% | ⚫ **Inviable / No Utilizado** |

**Insights Clave:**
1.  **Saturación Crítica:** La operación está al límite en **CD12** y **CD09**. Un aumento mínimo en la demanda de sus zonas asignadas haría el sistema infactible (o requeriría penalizaciones).
2.  **Inteligencia del Solver:** El modelo decidió **no utilizar el depósito CD06** (Capacidad: 3kg). Al detectar que su capacidad era insuficiente para las demandas cercanas (que rondan los 10-15kg), el algoritmo prefirió rutear vehículos desde depósitos más lejanos pero con capacidad (como CD05 o CD09), evitando así el sobrecosto por "multa" de capacidad.

### Conclusión y Recomendaciones
El modelo actual es **operativamente factible y matemáticamente óptimo**. Para mejorar la robustez de la cadena de suministro ante variaciones futuras de demanda, se recomienda:
1.  **Ampliar la capacidad del CD12:** Es el punto más frágil de la red.
2.  **Cerrar o Reestructurar CD06:** Su capacidad actual de 3kg es irrelevante para la magnitud de los pedidos (promedio 11kg/cliente), haciéndolo inútil en la práctica logística actual.

## Caso 3: Modelo Logístico Realista con Múltiples Centros y Vehículos Heterogéneos

### Evolución del Modelo

Este escenario representa la versión más completa y cercana a una **operación logística urbana real**.  
A diferencia de los Casos 1 y 2, aquí convergen simultáneamente tres dimensiones clave:

- **Múltiples Centros de Distribución (CDs)** con capacidades y ubicaciones diferenciadas.  
- **Vehículos heterogéneos**, cada uno con capacidad, costos y desempeño propios.  
- **Restricciones urbanas y operacionales**, incluyendo zonas exclusivas por tipo de vehículo, horarios de operación y limitaciones de acceso.

El desafío ya no consiste únicamente en trazar rutas eficientes, sino en lograr una **coordinación integral del sistema logístico**:  
determinar qué vehículo parte de qué centro, qué clientes puede atender según las restricciones urbanas y cómo **minimizar el costo logístico total** en un entorno complejo y dinámico.

### Enfoque de Resolución

Para manejar simultáneamente la escala del problema (50–100 clientes) y las restricciones heterogéneas, se adopta un **enfoque híbrido** que combina modelación matemática rigurosa con algoritmos heurísticos avanzados.

#### 1. Modelación Base con Pyomo
Se definen formalmente:
- los **conjuntos** (clientes, vehículos, depósitos),
- los **parámetros** (capacidades, demandas, costos, distancias),
- y las **restricciones** principales.

Esto garantiza la **correspondencia exacta** entre la formulación matemática (Etapa 1) y su implementación computacional.

#### 2. Resolución del VRP mediante OR-Tools
Debido al crecimiento cuadrático del número de arcos, un solver **MILP exacto** deja de ser escalable para este caso.  
OR-Tools permite:

- obtener soluciones **viables y de buena calidad**,  
- en tiempos computacionales razonables,  
- manteniendo la posibilidad de incorporar restricciones complejas del entorno urbano.

Su uso constituye un **trade-off** entre optimalidad teórica y viabilidad práctica.

#### 3. Restricciones Urbanas por Tipo de Vehículo
La **matriz de costos** se modifica introduciendo penalizaciones prohibitivas para impedir que un vehículo:

- ingrese a zonas donde no está autorizado,
- exceda horarios,
- o viole reglas de circulación urbana.

Este mecanismo garantiza la coherencia entre la operación real y la solución calculada.

#### 4. Capacidad y Posible Retorno al Centro
El modelo contempla que un vehículo pueda **regresar a su centro** para reabastecerse o descargar, lo cual refleja mejor la operación real y evita rutas físicamente inviables.

Este enfoque híbrido permite equilibrar:

- **factibilidad computacional**,  
- **realismo operativo**,  
- y **calidad de la solución**,

logrando un diseño robusto para apoyar decisiones empresariales en contextos logísticos urbanos complejos.


In [10]:
!pip install ortools folium pyomo

In [11]:

#  Configuración de rutas locales


# Solo cambiamos las rutas de entrada/salida
CLIENTS_CSV  = "/content/clients.csv"
DEPOTS_CSV   = "/content/depots.csv"
VEHICLES_CSV = "/content/vehicles.csv"
PARAM_CSV    = "/content/parameters_urban.csv"

OUTPUT_VERIF = "/content/verificacion_caso3.csv"
OUTPUT_MAP   = "/content/map_caso3.html"

print("Archivos de entrada:")
print(" -", CLIENTS_CSV)
print(" -", DEPOTS_CSV)
print(" -", VEHICLES_CSV)
print(" -", PARAM_CSV)
print("Archivos de salida:")
print(" -", OUTPUT_VERIF)
print(" -", OUTPUT_MAP)


Archivos de entrada:
 - /content/clients.csv
 - /content/depots.csv
 - /content/vehicles.csv
 - /content/parameters_urban.csv
Archivos de salida:
 - /content/verificacion_caso3.csv
 - /content/map_caso3.html


In [12]:

import pandas as pd
import math, csv, os
import folium

# OR-Tools
from ortools.constraint_solver import pywrapcp, routing_enums_pb2

# Pyomo mínimo (solo conjuntos y parámetros)
from pyomo.environ import ConcreteModel, Set, Param


# ---------------------------
# Distancia Haversine
# ---------------------------
def haversine_km(lat1, lon1, lat2, lon2):
    R = 6371.0
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlambda = math.radians(lon2 - lon1)
    a = (math.sin(dphi/2)**2 +
         math.cos(phi1)*math.cos(phi2)*math.sin(dlambda/2)**2)
    return 2 * R * math.atan2(math.sqrt(a), math.sqrt(1-a))


# ---------------------------
# Lectura de datos
# ---------------------------
clients  = pd.read_csv(CLIENTS_CSV)
depots   = pd.read_csv(DEPOTS_CSV)
vehicles = pd.read_csv(VEHICLES_CSV)
params   = pd.read_csv(PARAM_CSV)

clients["id"] = clients["StandardizedID"]
clients["lat"] = clients["Latitude"]
clients["lon"] = clients["Longitude"]
clients["demand"] = clients["Demand"]
clients["restriction"] = clients.get("VehicleSizeRestriction", None)

depots["id"] = depots["StandardizedID"]
depots["lat"] = depots["Latitude"]
depots["lon"] = depots["Longitude"]

vehicles["id"] = vehicles["StandardizedID"]
vehicles["capacity"] = vehicles["Capacity"]
vehicles["vtype"] = vehicles.get("VehicleType", "")


# ---------------------------
# Asignación de depósitos a vehículos
# ---------------------------
if "DepotID" in vehicles.columns:
    vehicles["start_depot"] = vehicles["DepotID"].astype(str)
    print("DepotID encontrado — se usa el archivo directamente.")
else:
    depot_ids = list(depots["id"])
    vehicles["start_depot"] = [ depot_ids[i % len(depot_ids)] for i in range(len(vehicles)) ]
    print("DepotID ausente — asignación automática round-robin.")


# ---------------------------
# Parámetros de costo
# ---------------------------
P = {row["Parameter"]: row["Value"] for _, row in params.iterrows()}
fuel_price = float(P.get("fuel_price", 16300))
fuel_eff   = float(P.get("fuel_efficiency_van_small_min", 30))
fixed_cost = float(P.get("C_fixed", 50000))
cost_km    = float(P.get("C_dist", 2500))
cost_hour  = float(P.get("C_time", 7600))
avg_speed  = 30.0


# ===============================================================
# PYOMO (mínimo) — Definición de conjuntos y parámetros solamente
# ===============================================================

model = ConcreteModel()

# conjunto de depósitos
model.D = Set(initialize=list(depots["id"]))

# conjunto de clientes
model.C = Set(initialize=list(clients["id"]))

# conjunto de vehículos
model.V = Set(initialize=list(vehicles["id"]))

# demanda del cliente
demand_dict = {row["id"]: float(row["demand"]) for _, row in clients.iterrows()}
model.d = Param(model.C, initialize=demand_dict)

# capacidad del vehículo
cap_dict = {row["id"]: float(row["capacity"]) for _, row in vehicles.iterrows()}
model.cap = Param(model.V, initialize=cap_dict)

print("Pyomo: conjuntos y parámetros cargados correctamente.")


# ===============================================================
# Construcción de nodos para OR-Tools
# ===============================================================

nodes = []
for _, d in depots.iterrows():
    nodes.append({"id": d["id"], "lat": float(d["lat"]), "lon": float(d["lon"]), "type": "depot"})

for _, c in clients.iterrows():
    nodes.append({
        "id": c["id"],
        "lat": float(c["lat"]),
        "lon": float(c["lon"]),
        "type": "client",
        "demand": float(c["demand"]),
        "restriction": c["restriction"]
    })

node_index = {n["id"]: i for i, n in enumerate(nodes)}
num_nodes = len(nodes)


# ---------------------------
# Matriz de distancias
# ---------------------------
dist_km = [[0]*num_nodes for _ in range(num_nodes)]
dist_m  = [[0]*num_nodes for _ in range(num_nodes)]

for i in range(num_nodes):
    for j in range(num_nodes):
        if i != j:
            d = haversine_km(nodes[i]["lat"], nodes[i]["lon"], nodes[j]["lat"], nodes[j]["lon"])
            dist_km[i][j] = d
            dist_m[i][j] = int(d * 1000)


# ===============================================================
# OR-TOOLS VRP
# ===============================================================

num_vehicles = len(vehicles)
starts = [ node_index[vehicles.iloc[v]["start_depot"]] for v in range(num_vehicles) ]
ends = starts[:]  # mismo depósito de retorno

manager = pywrapcp.RoutingIndexManager(num_nodes, num_vehicles, starts, ends)
routing = pywrapcp.RoutingModel(manager)


# ---------------------------
# Callback de distancia con restricciones por tamaño de vehículo
# ---------------------------
def make_dist_cb(v_idx):
    vtype = str(vehicles.iloc[v_idx]["vtype"]).strip().lower()
    def cb(from_index, to_index):
        fn = manager.IndexToNode(from_index)
        tn = manager.IndexToNode(to_index)
        if nodes[tn]["type"] == "client":
            need = nodes[tn]["restriction"]
            if need is not None and str(need).strip() != "":
                if str(need).strip().lower() != vtype:
                    return int(1e9)
        return dist_m[fn][tn]
    return cb

for v in range(num_vehicles):
    idx = routing.RegisterTransitCallback(make_dist_cb(v))
    routing.SetArcCostEvaluatorOfVehicle(idx, v)


# ---------------------------
# Capacidad
# ---------------------------
demands = [0]*num_nodes
for i,n in enumerate(nodes):
    if n["type"] == "client":
        demands[i] = int(n["demand"])

def demand_cb(index):
    return demands[ manager.IndexToNode(index) ]

demand_idx = routing.RegisterUnaryTransitCallback(demand_cb)
vehicle_caps = [int(c) for c in list(vehicles["capacity"])]
routing.AddDimensionWithVehicleCapacity(demand_idx, 0, vehicle_caps, True, "Capacity")


# ---------------------------
# Solución
# ---------------------------
params = pywrapcp.DefaultRoutingSearchParameters()
params.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC
params.local_search_metaheuristic = routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH
params.time_limit.seconds = 30

solution = routing.SolveWithParameters(params)
print("Solución encontrada.")


# ===============================================================
# Archivo de verificación
# ===============================================================
with open(OUTPUT_VERIF, "w", newline='', encoding="utf-8") as f:
    w = csv.writer(f)
    w.writerow(["VehicleId","DepotId","InitialLoad","RouteSequence","ClientsServed",
                "DemandsSatisfied","TotalDistance","TotalTime","FuelCost"])

    for v in range(num_vehicles):
        start_idx = routing.Start(v)
        idx = start_idx
        route = []
        dem_list = []
        total_dist = 0
        total_time = 0
        clients_served = 0
        load = 0

        while not routing.IsEnd(idx):
            n = manager.IndexToNode(idx)
            route.append(nodes[n]["id"])

            nxt = solution.Value(routing.NextVar(idx))
            nxt_node = manager.IndexToNode(nxt)
            total_dist += dist_km[n][nxt_node]

            if nodes[n]["type"] == "client":
                d = nodes[n]["demand"]
                dem_list.append(int(d))
                load += d
                clients_served += 1

            total_time += (dist_km[n][nxt_node] / avg_speed) * 60
            idx = nxt

        n = manager.IndexToNode(idx)
        route.append(nodes[n]["id"])

        gallons = total_dist / fuel_eff
        fuel_cost = gallons * fuel_price

        if clients_served == 0:
            w.writerow([vehicles.iloc[v]["id"], vehicles.iloc[v]["start_depot"], 0,
                        vehicles.iloc[v]["start_depot"]+"-"+vehicles.iloc[v]["start_depot"],
                        0,"",0,0,0])
        else:
            w.writerow([
                vehicles.iloc[v]["id"],
                vehicles.iloc[v]["start_depot"],
                round(load,2),
                "-".join(route),
                clients_served,
                "-".join(str(x) for x in dem_list),
                round(total_dist,3),
                round(total_time,2),
                round(fuel_cost,2)
            ])

print("Archivo verificacion_caso3.csv escrito:", OUTPUT_VERIF)

# Mapa Folium
m = folium.Map(location=[4.70, -74.10], zoom_start=11)

for _, d in depots.iterrows():
    folium.Marker([d["lat"], d["lon"]], popup=f"Depósito: {d['id']}",
                  icon=folium.Icon(color="red")).add_to(m)

for _, c in clients.iterrows():
    folium.CircleMarker([c["lat"], c["lon"]], radius=4,
                         popup=f"Cliente: {c['id']}<br>Demanda: {c['demand']}",
                         color="blue", fill=True).add_to(m)

colors = ["red","blue","green","purple","orange","black","darkgreen","cadetblue","pink"]

for v in range(num_vehicles):
    idx = routing.Start(v)
    coords = []
    while not routing.IsEnd(idx):
        n = manager.IndexToNode(idx)
        coords.append((nodes[n]["lat"], nodes[n]["lon"]))
        idx = solution.Value(routing.NextVar(idx))
    n = manager.IndexToNode(idx)
    coords.append((nodes[n]["lat"], nodes[n]["lon"]))
    if len(coords) > 2:
        folium.PolyLine(coords, color=colors[v % len(colors)], weight=3).add_to(m)

m.save(OUTPUT_MAP)
print("Mapa generado:", OUTPUT_MAP)


DepotID ausente — asignación automática round-robin.
Pyomo: conjuntos y parámetros cargados correctamente.
Solución encontrada.
Archivo verificacion_caso3.csv escrito: /content/verificacion_caso3.csv
Mapa generado: /content/map_caso3.html


In [13]:
display(m)

In [14]:
import pandas as pd
pd.read_csv(OUTPUT_VERIF).head()

VehicleId DepotId  InitialLoad  \
0      V001    CD01        121.0   
1      V002    CD02          0.0   
2      V003    CD03          0.0   
3      V004    CD04          9.0   
4      V005    CD05          0.0   

                                       RouteSequence  ClientsServed  \
0  CD01-C004-C047-C002-C063-C081-C082-C012-C020-C...              9   
1                                          CD02-CD02              0   
2                                          CD03-CD03              0   
3                                     CD04-C007-CD04              1   
4                                          CD05-CD05              0   

             DemandsSatisfied  TotalDistance  TotalTime  FuelCost  
0  22-12-15-12-12-12-12-12-12         25.447      50.89  11850.92  
1                         NaN          0.000       0.00      0.00  
2                         NaN          0.000       0.00      0.00  
3                           9          0.666       1.33    310.01  
4                         NaN          0.000       0.00      0.00

### Análisis Estratégico del Caso 3

#### 1. Parámetros Críticos que Afectan la Operación

Del análisis de sensibilidad se identifican tres parámetros con **impacto dominante** sobre la operación logística:

- **Costo de combustible**: influye directamente en el costo total y favorece soluciones con rutas más cortas y eficientes.  
- **Demandas de los clientes**: determinan el nivel de saturación del sistema y condicionan el número de vehículos necesarios.  
- **Capacidad de los Centros de Distribución (CDs)**: modifica la congestión operativa, la longitud de las rutas y la eficiencia global del sistema.

Estos parámetros redefinen la **estructura del VRP** y deben ser monitoreados continuamente en la planificación operativa diaria.


#### 2. Cuellos de Botella Operacionales

El análisis revela tres fuentes principales de congestión:

- **Centros de distribución saturados**, con más demanda asignada de la que su infraestructura puede manejar.  
- **Zonas urbanas restringidas**, donde únicamente pueden operar vehículos pequeños.  
- **Clientes geográficamente dispersos**, que generan rutas largas y tiempos de operación elevados.

Detectar estos cuellos de botella permite **priorizar inversiones**, rediseñar rutas y optimizar el uso de la flota.


#### 3. Impacto de las Variaciones de Parámetros

- **Costo de combustible (±20 %)**: modifica significativamente el costo total de operación; el impacto es mayor en vehículos con baja eficiencia energética.  
- **Capacidad de los centros**: aumentar la capacidad reduce rutas redundantes y mejora el desempeño logístico general.  
- **Demandas de los clientes**: incrementos fuertes generan necesidades adicionales de reabastecimiento y mayores requerimientos de flota.

Estos factores muestran que el sistema es **sensible tanto a aspectos operativos como energéticos**.


#### 4. Recomendaciones para LogistiCo

A partir de los resultados, se sugieren las siguientes acciones estratégicas:

- **Incrementar la capacidad** o mejorar la distribución de carga entre los centros existentes.  
- **Refinar la flota**, fortaleciendo la disponibilidad de vehículos pequeños en zonas con restricciones urbanas.  
- **Incorporar modelos predictivos** que anticipen fluctuaciones en la demanda y ajusten la operación.  
- **Monitorear continuamente el costo del combustible** para adaptar las políticas de asignación y despliegue vehicular.

Estas medidas contribuyen a una operación más estable, flexible y resiliente.

### 5. Preguntas Estratégicas del Caso

**¿Cuáles parámetros influyen más en la logística urbana?**  
Principalmente el **costo del combustible**, las **demandas** y la **capacidad** de los centros.

**¿Dónde se presentan los mayores cuellos de botella?**  
En **centros saturados** y en **zonas con restricciones operativas**.

**¿Qué mejoras se recomiendan?**  
Una **flota más diversificada**, una **redistribución geográfica de carga** y una **expansión selectiva de centros críticos**.
